racket/collects/json/main.rkt

```diff
@@ -166,49 +187,56 @@
           [(eq? x #f)     (write-bytes #"false" o)]
           [(eq? x #t)     (write-bytes #"true" o)]
           [(eq? x jsnull) (write-bytes #"null" o)]
-          [(string? x) (write-json-string x)]
-          [(list? x)
-           (write-bytes #"[" o)
-           (when (pair? x)
-             (for/fold ([first? #t])
-                       ([x (in-list x)])
-               (unless first? (write-bytes #"," o))
-               (format/write-indented-newline)
-               (format/write-indent-bytes)
-               (write-jsval x (add1 layer))
-               #f)
-             (format/write-indented-newline))
-           (write-bytes #"]" o)]
-          [(hash? x)
+          [(string-rep? x) (write-json-string (string-rep->string x))]
+          [(list-rep? x)
+           (let ([x (list-rep->list x)])
+             (write-bytes #"[" o)
+             (when (pair? x)
+               (for/fold ([first? #t])
+                         ([x (in-list x)])
+                 (unless first? (write-bytes #"," o))
+                 (format/write-indented-newline)
+                 (format/write-indent-bytes)
+                 (write-jsval x (add1 layer))
+                 #f)
+               (format/write-indented-newline))
+             (write-bytes #"]" o))]
+          [(object-rep? x)
            (define write-hash-kv
              (let ([first? #t])
                (λ (k v)
-                 (unless (symbol? k)
+                 (unless (key-rep? k)
                    (raise-type-error who "legal JSON key value" k))
                  (if first? (set! first? #f) (write-bytes #"," o))
                  (format/write-indented-newline)
                  (format/write-indent-bytes)
                  ;; use a string encoding so we get the same deal with
                  ;; `rx-to-encode'
-                 (write-json-string (symbol->immutable-string k))
+                 (write-json-string (key-rep->string k))
                  (write-bytes #":" o)
                  (format/write-whitespace)
                  (write-jsval v (add1 layer)))))
-           (write-bytes #"{" o)
-           (unless (hash-empty? x)
-             (hash-for-each x write-hash-kv #t)
-             (format/write-indented-newline))
-           (write-bytes #"}" o)]
+           (let ([x (object-rep->hash x)])
+             (write-bytes #"{" o)
+             (unless (hash-empty? x)
+               (hash-for-each x write-hash-kv #t)
+               (format/write-indented-newline))
+             (write-bytes #"}" o))]
           [else (raise-type-error who "legal JSON value" x)]))
   (void))
 
 ;; -----------------------------------------------------------------------------
 ;; PARSING (from JSON to Racket)
 
 (define (read-json [i (current-input-port)] #:null [jsnull (json-null)])
-  (read-json* 'read-json i jsnull))
+  (read-json* 'read-json i jsnull make-immutable-hasheq values string->symbol values))
```

Comment:
Would it be possible to add a `#:mhash?` keyword to `read-json` to allow users to create mutable hashtables for JSON objects? This would be helpful for scenarios where JSON objects need to be dynamically modified (e.g., updating keys in nested objects) while maintaining backward compatibility with the default immutable hashtables. Since `jsexpr?` already considers mutable hashtables as valid JSON objects, this addition seems like a natural extension.

Reply:
Hi @NoahStoryM,

That is out of the scope of this PR.  But (if this PR is accepted) it may be easier to propose as a new PR.

Cheers,
Sam